### ISSUE TITLE
git install doesn't recognize new `main` default branch

### ISSUE THREAD

**Issue by Wasdee** [python-poetry/poetry#3366]
`poetry add git+https://github.com/AGeekInside/fastapi-jinja.git` result in exception since it is tires to checkout `master` branch.

```
poetry add git+https://github.com/AGeekInside/fastapi-jinja.git

Updating dependencies
Resolving dependencies...

  CalledProcessError

  Command '['git', '--git-dir', 'C:/Users/n2/AppData/Local/Temp/pypoetry-git-fastapi-jinjau3blv69e/.git', '--work-tree', 'C:/Users/n2/AppData/Local/Temp/pypoetry-git-fastapi-jinjau3blv69e', 'checkout', 'master']' returned non-zero exit status 1.

  at ~\.poetry\lib\poetry\utils\_compat.py:217 in run
      213│                 process.wait()
      214│                 raise
      215│             retcode = process.poll()
      216│             if check and retcode:
    → 217│                 raise CalledProcessError(
      218│                     retcode, process.args, output=stdout, stderr=stderr
      219│                 )
      220│         finally:
      221│             # None because our context manager __exit__ does not use them.

```

as the default branch name has changed, poetry should both accept `main` and `master`. 

the workaround is 
1. `poetry add git+https://github.com/AGeekInside/fastapi-jinja.git#main`
2. `pip install git+https://github.com/AGeekInside/fastapi-jinja.git`

**Comment by CircleOnCircles** [comment]
I could create a PR if you like.

**Comment by sinoroc** [comment]
Based on what clue should it be decided to go for `master` or `main`?

>  the default branch name has changed

Where? In git, in github, in your own projects?

Can the default branch name be auto-detected somehow?

**Comment by CircleOnCircles** [comment]
@sinoroc github has changed for all newly created repo. https://www.zdnet.com/article/github-to-replace-master-with-main-starting-next-month/

My impulse answer would be to check (or to try-catch) starting for 'master' then 'main'. 

or, we can derive the implementation from pip.

**Comment by CircleOnCircles** [comment]
One thing to note also that, the verbose message of the error is not really helpful if not looking carefully. The verbose message should tell the user about the cause that the target branch is not exist.

**Comment by abn** [comment]
Poetry should simply not specify a branch by default. This should then default to whatever is the server default and it should adhere to expected behaviour. Might require removing this "master is default" assumption at multiple points in the code. PR welcome.

As for the error messages. There is very little we can do here at the moment since the  `git` command is being processed as a subprocess call. Moving to a native python library will help with this in the future. As for immideate improvements, if you can identify any easy wins within the current error handling, PR is welcome, but maybe open another issue for that.

**Comment by sinoroc** [comment]
The only meaningful direct reference in code I could find is [there](https://github.com/python-poetry/poetry-core/blob/c7a4c31684f00e7f3a677d922b73fbead91528d2/poetry/core/packages/vcs_dependency.py#L32-L34):

```python
        if not any([branch, tag, rev]):
            # If nothing has been specified, we assume master
            branch = "master"
```

The rest of the code should probably be adapted to work when `branch`, `tag` and `rev` are all `None`.

**Comment by Armaxxx** [comment]
Is this still open? I could try to tackle this as my first issue.

**Comment by sinoroc** [comment]
> Is this still open? I could try to tackle this as my first issue.

Yes, as far as I can tell, there is no PR yet so you'd be first.

**Comment by Armaxxx** [comment]
This is definitely a dumb question, but I need to work on poetry-core, right? Not in the poetry repo.

**Comment by sinoroc** [comment]
> I need to work on poetry-core, right? Not in the poetry repo.

As far as I can tell, yes. At least, I would start with _poetry-core_. Then it will show if some more modifications are needed in _poetry_ or not.

**Comment by TalyOwn** [comment]
I've got the same error with master branch, which is default in my repository. So, in pyprojeft.toml
```
common = { git = "git@github.com:talycraft/talywalker-common.git", branch = "master" }
```
or
```
common = { git = "https://github.com/talycraft/talywalker-common.git", branch = "master" }
```
both variants leads to
```
Step 14/20 : RUN     poetry config virtualenvs.create false &&     poetry install --no-interaction --no-ansi
 ---> Running in 8754c1bfa0d8
Skipping virtualenv creation, as specified in config file.
Installing dependencies from lock file
Warning: The lock file is not up to date with the latest changes in pyproject.toml. You may be getting outdated dependencies. Run update to update them.

  CalledProcessError

  Command '['git', 'clone', '--recurse-submodules', 'git@github.com:talycraft/talywalker-common.git', '/tmp/pypoetry-git-talywalker-commonaa864cw7']' returned non-zero exit status 128.

  at ~/.poetry/lib/poetry/utils/_compat.py:217 in run
      213│                 process.wait()
      214│                 raise
      215│             retcode = process.poll()
      216│             if check and retcode:
    → 217│                 raise CalledProcessError(
      218│                     retcode, process.args, output=stdout, stderr=stderr
      219│                 )
      220│         finally:
      221│             # None because our context manager __exit__ does not use them.
The command '/bin/sh -c poetry config virtualenvs.create false &&     poetry install --no-interaction --no-ansi' returned a non-zero code: 1
```

**Comment by Shams66789** [comment]
Is it still open ? I would like to take it as my first issue.

**Comment by CircleOnCircles** [comment]
> Is it still open ? I would like to take it as my first issue.

You might want to take a look here first. https://github.com/python-poetry/poetry-core/pull/117

**Comment by abn** [comment]
Reopening as the PR was reverted.

**Comment by marcoaaguiar** [comment]
What is the status of this? It was opened and closed, merged and reverted, so many times that I got lost 😟

**Comment by lambdalisue** [comment]
I still faced this issue on the latest Poetry (1.1.6)

**Comment by lambdalisue** [comment]
I still faced this issue on the latest Poetry (1.1.7)...

**Comment by grimmer0125** [comment]
I have submitted a PR to solve this, #4386.

**Comment by joooeey** [comment]
Here's the set of steps to reproduce the same issue occuring when I edit the `pyproject.toml` file. Note that #4386 as it is now would not solve the problem in my case because it relies on editing `pyproject.toml`. I think this should be solved further upstream.

# Steps:

0. `cd ~/Desktop`
1. `poetry new poetry-demo`
2. Edit `pyproject.toml` to include:
```
[tool.poetry.dependencies]
python = "^3.8"  # Leave this line as it is.
sentinelsat = { git = "https://github.com/sentinelsat/sentinelsat.git" }
```
3. `cd poetry-demo`
4. `poetry install`

resutls in:

```
  CalledProcessError

  Command '['git', '--git-dir', '/tmp/pypoetry-git-sentinelsao21ujftz/.git', '--work-tree', '/tmp/pypoetry-git-sentinelsao21ujftz', 'checkout', 'master']' returned non-zero exit status 1.

  at ~/.poetry/lib/poetry/utils/_compat.py:217 in run
      213│                 process.wait()
      214│                 raise
      215│             retcode = process.poll()
      216│             if check and retcode:
    → 217│                 raise CalledProcessError(
      218│                     retcode, process.args, output=stdout, stderr=stderr
      219│                 )
      220│         finally:
      221│             # None because our context manager __exit__ does not use them.

```

# Workaround:

Use `sentinelsat = { git = "https://github.com/sentinelsat/sentinelsat.git", branch = "main" }`. This is the workaround that #4386 uses. However, if possible it would be better to address this deeper inside Poetry.

**Comment by hugolytics** [comment]
My poetry packages are not installable from git unless i specify the #main branch.
Is this going to be resolved or should i always install this way?

**Comment by finswimmer** [comment]
It is fixed by https://github.com/python-poetry/poetry-core/pull/192 which is available in poetry-core 1.1.0a7. So this will be fixed by (re)intalling the latest poetry pre-release version (1.2.0a2).
